### PR TITLE
643 tutorial for custom backbone

### LINKED ISSUE #643: Tutorial for custom backbone
## Subject of the documentation

Add a tutorial for applying a custom backbone with registy.

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

- Add custom backbone.
- Add backbone interface.
- Some refactoring of code.
- Some fixes to the docs.

Fixes #643

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [x] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
@YipengHu @zacbaum @s-sd  would you have time for review?

- updated registry doc  https://deepreg--644.org.readthedocs.build/en/644/docs/registry.html
- example script https://github.com/DeepRegNet/DeepReg/blob/643-tutorial-for-custom-backbone/examples/custom_backbone.py.

File: docs/source/docs/registry.md
Comment (YipengHu) [inline docs/source/docs/registry.md:46]: at this point, it is not clear what a "configuration" is and what a "config file" is - better to clarify that this is the ones used for DeepReg and perhaps a link to the config in the doc.

File: docs/source/docs/registry.md
Comment (YipengHu) [inline docs/source/docs/registry.md:67]: where are the examples for each category?

File: examples/README.md
Comment (YipengHu) [inline examples/README.md:4]: certain -> customising (this is the only one we are showcasing for now, if others, we can change the wording here)
like -> such as (a bit more formal)

**Review by YipengHu — CHANGES_REQUESTED** [review]

File: docs/source/docs/registry.md
Comment (mathpluscode) [reply docs/source/docs/registry.md:46]: I think the second phrase is not necessary, a link of config doc above should solve the issue.

File: docs/source/docs/registry.md
Comment (mathpluscode) [reply docs/source/docs/registry.md:67]: as following, "custom backbone", no?

**Comment by mathpluscode** [comment]
@YipengHu please have a second look ^^ the registered classes have been resumed here https://deepreg--644.org.readthedocs.build/en/644/docs/registered_classes.html

This page will be automatically generated and checked in unit test

File: docs/source/docs/registry.md
Comment (mathpluscode) [reply docs/source/docs/registry.md:70]: https://github.com/DeepRegNet/DeepReg/blob/main/examples/custom_backbone.py

Link will be here, but need to wait until merged.

File: docs/source/docs/registry.md
Comment (s-sd) [inline docs/source/docs/registry.md:32]: Consider changing to: 'The decorator automatically registers the class upon import. To ensure that this class is registered when `import deepreg` is called, this class and related parent modules need to be imported in the `__init__.py` files.'

File: docs/source/docs/registry.md
Comment (s-sd) [inline docs/source/docs/registry.md:64]: Is this supposed to be `build_backbone`? Change to:
`unet = REGISTRY.build_backbone(config={"name":"unet", **kwargs})`

File: docs/source/docs/registry.md
Comment (s-sd) [reply docs/source/docs/registry.md:70]: I think maybe consider changing to: 'A detailed example is provided for the `BACKBONE_CLASS` category below' or something similar.

Using 'each category' could be understood as meaning that you have a detailed example for each category i.e. `BACKBONE_CLASS`, `LOSS_CLASS`, `MODEL_CLASS`, `DATA_AUGMENTATION_CLASS` so I think it is best to just specify or change 'each category' to 'some categories' for this sentence.

**Review by s-sd — COMMENTED** [review]
Looks great @mathpluscode! Added a few comments

File: docs/source/docs/registry.md
Comment (mathpluscode) [reply docs/source/docs/registry.md:70]: 😂oh we should have examples per category, just need to do it.. I see why it's confusing now. Will change.

**Comment by mathpluscode** [comment]
@YipengHu @s-sd I've addressed your comments :) have another look plz

File: docs/source/docs/registry.md
Comment (YipengHu) [inline docs/source/docs/registry.md:101]: - better to use "subclassing" instead of "inherit" to be consistent with other pytorch/keras language?
- Is the class name 'BackboneInterface' confusing? as your category is just 'backbone_class', it is not obvious what the rest of the class names are, "ModelInterface" and "LossInterface"? better to be just 'Backbone', 'Loss', 'Model'....

Comment (mathpluscode) [reply docs/source/docs/registry.md:101]: @YipengHu ok

Comment (mathpluscode) [reply docs/source/docs/registry.md]: done

File: docs/source/docs/registered_classes.md
Comment (YipengHu) [inline docs/source/docs/registered_classes.md:47]: i would: ddf -> nonrigid

File: docs/source/docs/registered_classes.md
Comment (YipengHu) [inline docs/source/docs/registered_classes.md:15]: hide the "model" category now, it would be easier to add later - not modifying.

Comment (mathpluscode) [reply docs/source/docs/registered_classes.md:15]: not here, this file is generated automatically.

File: docs/source/docs/registered_classes.md
Comment (mathpluscode) [reply docs/source/docs/registered_classes.md:47]: it is applying ddf, so i dont see the point to be less specific

Comment (YipengHu) [reply docs/source/docs/registered_classes.md:47]: strictly speaking, affine use ddf to warp images too

**Review by YipengHu — APPROVED** [review]

**Review by s-sd — APPROVED** [review]

**Pull request merged**
